url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/internal/chartpart/Axis.java

```diff
@@ -0,0 +1,421 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.knowm.xchart.internal.chartpart;
+
+import java.awt.Graphics2D;
+import java.awt.Shape;
+import java.awt.font.FontRenderContext;
+import java.awt.font.TextLayout;
+import java.awt.geom.AffineTransform;
+import java.awt.geom.Rectangle2D;
+import java.util.List;
+
+import org.knowm.xchart.internal.Series;
+import org.knowm.xchart.internal.Series_AxesChart;
+import org.knowm.xchart.style.AxesChartStyler;
+import org.knowm.xchart.style.CategoryStyler;
+import org.knowm.xchart.style.Styler.LegendPosition;
+
+/**
+ * Axis
+ *
+ * @author timmolter
+ */
+public class Axis<ST extends AxesChartStyler, S extends Series> implements ChartPart {
+
+  public enum AxisDataType {
+
+    Number, Date, String;
+  }
+
+  private final Chart<AxesChartStyler, Series_AxesChart> chart;
+  private Rectangle2D bounds;
+
+  private final AxesChartStyler stylerAxesChart;
+
+  /** the axisDataType */
+  private AxisDataType axisDataType;
+
+  /** the axis title */
+  private AxisTitle<AxesChartStyler, Series_AxesChart> axisTitle;
+
+  /** the axis tick */
+  private AxisTick<AxesChartStyler, Series_AxesChart> axisTick;
+
+  /** the axis tick calculator */
+  private AxisTickCalculator_ axisTickCalculator;
+
+  /** the axis direction */
+  private Direction direction;
+
+  private double min;
+
+  private double max;
+
+  /** An axis direction */
+  public enum Direction {
+
+    /** the constant to represent X axis */
+    X,
+
+    /** the constant to represent Y axis */
+    Y
+  }
+
+  /**
+   * Constructor
+   *
+   * @param chart the Chart
+   * @param direction the axis direction (X or Y)
+   */
+  public Axis(Chart<AxesChartStyler, Series_AxesChart> chart, Direction direction) {
+
+    this.chart = chart;
+    this.stylerAxesChart = chart.getStyler();
+
+    this.direction = direction;
+    axisTitle = new AxisTitle<AxesChartStyler, Series_AxesChart>(chart, direction);
+    axisTick = new AxisTick<AxesChartStyler, Series_AxesChart>(chart, direction);
+  }
+
+  /**
+   * Reset the default min and max values in preparation for calculating the actual min and max
+   */
+  protected void resetMinMax() {
+
+    min = Double.MAX_VALUE;
+    max = -Double.MAX_VALUE;
+  }
+
+  /**
+   * @param min
+   * @param max
+   */
+  protected void addMinMax(double min, double max) {
+
+    // System.out.println(min);
+    // System.out.println(max);
+    // NaN indicates String axis data, so min and max play no role
+    if (this.min == Double.NaN || min < this.min) {
+      this.min = min;
+    }
+    if (this.max == Double.NaN || max > this.max) {
+      this.max = max;
+    }
+
+    // System.out.println(this.min);
+    // System.out.println(this.max);
+  }
+
+  @Override
+  public void paint(Graphics2D g) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.